src/pyodide.js

```diff
@@ -171,102 +161,127 @@ var languagePluginLoader = new Promise((resolve, reject) => {
       return baseURL + path;
     };
 
-    let promise = new Promise((resolve, reject) => {
-      if (Object.keys(toLoad).length === 0) {
-        resolve('No new packages to load');
-        return;
+    if (Object.keys(toLoad).length === 0) {
+      return Promise.resolve('No new packages to load');
+    } else {
+      console.log(`Loading ${Object.keys(toLoad).join(', ')}`)
+    }
+
+    // When we load a package, we first load a JS file, and then the JS file
+    // loads the data file. The status of the first step is tracked by the
+    // loadScript function above. The status of the second step is tracked by
+    // emscripten's runDependency system.
+    //
+    // When the JS file is loaded, it *synchronously* adds a runDependency to
+    // emscripten. When it finishes loading, it removes the runDependency.
+    // Whenever the number of runDependencies changes, monitorRunDependency is
+    // called with the number of pending dependencies.
+    //
+    // Under this system, packages have completed loading if all javascript
+    // files are loaded, and then the number of runDependencies becomes zero.
+    // The number of runDependencies may happen to equal zero in between
+    // package files loading if we are loading multiple packages. We must avoid
+    // accidentally returning early if this happens.
+
+    // runDepPromise is a promise such that at any point in time, it is
+    // resolved if and only if the number of pending run dependencies is 0.
+    // We assume that at this point in time, there are no pending run
+    // dependencies.
+    //
+    // The promise it refers to may be replaced when a runDependency is added,
+    // so we must be careful to only await for this after we know all
+    // runDependencies have been added.
+    let runDepPromise = Promise.resolve();
+    // runDepResolve is the resolve function of runDepPromise if it is
+    // unresolved, undefined otherwise.
+    let runDepResolve = undefined;
+
+    // We must start monitoring before we load any script, as this function is
+    // triggered by the script loading.
+    self.pyodide._module.monitorRunDependencies = (n) => {
+      if (n === 0) {
+        if (runDepResolve !== undefined) {
+          runDepResolve();
+          runDepResolve = undefined;
+        }
+      } else {
+        if (runDepResolve === undefined) {
+          runDepPromise = new Promise(r => runDepResolve = r);
+        }
       }
+    };
 
-      let packageList = Array.from(Object.keys(toLoad));
-      _messageCallback(`Loading ${packageList.join(', ')}`)
+    // Add a promise that rejects when any exceptions is thrown in the process
+    // of loading a script. The promise never resolves, and we combine it
+    // with other promises via Promise.race.
+    let windowErrorHandler;
+    let windowErrorPromise = new Promise((_res, rej) => {
+      windowErrorHandler = e => rej(e.message);
+      self.addEventListener('error', windowErrorHandler);
+    });
```

Comment:
Do you know what this `windowErrorHandler` is supposed to handle? I think unless we can find a case where this gets triggered we should remove it. It seems dangerous to me: it could swallow error messages and respond to errors that are not actually related to loading the package. I would suggest removing it and if that causes trouble trying resolving the resulting problems in a less dangerous way.

Reply:
We load a package by loading an external javascript file via inserting a `<script>` tag. There is no reasonable way to catch errors from that file without blanket catching all errors. One way we can trigger this is by removing the `.data` file, but that's unpleasant to do in a test environment